docs/source/design.rst

```diff
@@ -1,7 +1,201 @@
+:class:`~ibis.expr.types.TableExpr`.
+
+Here's an example of each type of expression:
+
+.. code-block:: python
+
+   >>> import ibis
+   >>> t = ibis.table([('a', 'int64')])
+   >>> int64_column = t.a
+   >>> type(int64_column)
+   <class 'ibis.expr.types.Int64Column'>
+   >>> string_scalar = ibis.literal('some_string_value')
+   >>> type(string_scalar)
+   <class 'ibis.expr.types.StringScalar'>
+   >>> table_expr = t.mutate(b=t.a + 1)
+   >>> type(table_expr)
+   <class 'ibis.expr.types.TableExpr'>
+
+.. _node_class:
+
+The :class:`~ibis.expr.types.Node` Class
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+:class:`~ibis.expr.types.Node` subclasses make up the core set of operations of
+ibis. Each node corresponds to a particular operation.
+
+Most nodes are defined in the :mod:`~ibis.expr.operations` module.
+
+Examples of nodes include :class:`~ibis.expr.operations.Add` and
+:class:`~ibis.expr.operations.Sum`.
+
+Nodes have two important members (and often these are the only members defined):
+
+#. ``input_type``: a list of rules
+#. ``output_type``: a rule or method
+
+The ``input_type`` member is a list of rules that defines the types of
+the inputs to the operation. This is sometimes called the signature.
+
+The ``output_type`` member is a rule or a method that defines the output type
+of the operation. This is sometimes called the return type.
+
+An example of ``input_type``/``output_type`` usage is the
+:class:`~ibis.expr.operations.Log` class:
+
+.. code-block:: python
+
+   class Log(Node):
+
+       input_type = [
+           rules.double(),
+           rules.double(name='base', optional=True)
+       ]
+       output_type = rules.shape_like_arg(0, 'double')
+
+This class describes an operation called ``Log`` that takes one required
+argument: a double scalar or column, and one optional argument: a double scalar
+or column named ``base`` that defaults to nothing if not provided. The base
+argument is ``None`` by default so that the expression will behave as the
+underlying database does.
+
+These objects are instantiated when you use ibis APIs:
+
+.. code-block:: python
+
+   >>> import ibis
+   >>> t = ibis.table([('a', 'double')])
+   >>> log_1p = (1 + t.a).log()  # an Add and a Log are instantiated here
+
+.. _expr_vs_ops:
+
+Expressions vs Operations: Why are they different?
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Separating expressions from their underlying operations makes it easy to
+generically describe and validate the inputs to particular nodes. In the log
+example, it doesn't matter what *operation* (node) the double-valued arguments
+are coming from, they must only satisfy the requirement denoted by the rule.
+
+Separation of the :class:`~ibis.expr.types.Node` and
+:class:`~ibis.expr.types.Expr` classes also allows the API to be tied to the
+physical type of the expression rather than the particular operation, making it
+easy to define the API in terms of types rather than specific operations.
```

Comment:
Another key point here is that the operator output type will often depend on the input type(s). So the "user API" for the result of an operation is a strongly-typed expression having only the behavior of the actual output type of the operator.

* Add strings -> get strings, do things with strings
* Add numbers -> get numbers, do things with numbers

Reply:
Cool, added a blurb about this.